README.md
# python-halftone

A Python 3 module that uses [Pillow][pillow] to give images a halftone effect (see below for examples).

It is adapted from [this StackOverflow answer][so] (which includes example images) by [fraxel][fr].

Running it over large images, or with settings that create large images, can take some time.

Also see [Clay Flannigan's halftone][clay] in case that suits your needs better.

NOTE: Colors in the output files are darker in JPEG files compared to PNGs. This is something to do
with color profiles and RGB/CMYK conversion that I don't understand.
[Suggestions welcome on this Issue.][bug]

[pillow]: http://pillow.readthedocs.io
[so]: http://stackoverflow.com/questions/10572274/halftone-images-in-python/10575940#10575940
[fr]: http://stackoverflow.com/users/1175101/fraxel
[clay]: https://github.com/ClayFlannigan/halftone
[bug]: https://github.com/philgyford/python-halftone/issues/7


## Basic usage

```python
import halftone

h = halftone.Halftone("/path/to/myimage.jpg")
h.make()
```

Will create a new image at `/path/to/myimage_halftoned.jpg`, using the default settings.


## Options

There are a number of options that can be added to the `make()` call, e.g.:

```python
h.make(filename_addition="_new", scale=2)
```

The full list of options:

### `angles`

A list of angles, in degrees, that each channel should be rotated by. If `style="color"` then 4 angles, one per channel (CMYK, in that order) are required. If `style="grayscale"` only 1 angle is required; any extra are ignored. Experimenting with different angles can increase or reduce moiré patterns. [More on screen angles.](http://the-print-guide.blogspot.co.uk/2009/05/halftone-screen-angles.html)

Default: `[0, 15, 30, 45]`

### `antialias`

A boolean value, indicating whether the circles drawn should be antialiased. Because Pillow doesn't draw antialias shapes, this is done by drawing the image at 4x the size and then reducing it to the desired size, antialiasing as part of the process.

Default: `False`

### `filename_addition`

When saving the new image, this string will be added to the original filename. e.g. if the original filename is `"puppy.jpg"` and `filename_addition="_new"`, the saved file will be `"puppy_new.jpg"`.

Default: `"_halftoned"`

### `output_format`

Either `"default"`, `"jpeg"`, or `"png"`. What format should the halftone file be saved as? If `"default"` then it's the same as the original input image's format.

Default: `"default"`

### `output_quality`

When saving any JPEG images, what quality to use. From `0` to `100`. Pillow says to avoid anything over `95`. Has no effect on non-JPEG images.

Default: `75`

### `percentage`

How much of the gray component to remove from the CMY channels and put in the K channel.

Default: `0`

### `sample`

When creating each circle in the new image, what area of pixels should that circle represent?

Default: `10`

### `save_channels`

Whether to save the four CMYK channels as separate images, in addition to the combined halftoned image. Boolean. The files will have the color letter appended to the filename, like `puppy_halftoned_c.jpg` for the cyan channel. Only functions if the overall `style` argument is `"color"`.

Default: `False`

### `save_channels_format`

Either `"default"`, `"jpeg"`, or `"png"`. If `save_channels` is `True` then what format should the four separate images be saved as? If `"default"` then it's the same as the original input image's format.

Default: `"default"`

### `save_channels_style`

Either `"color"` or `"grayscale"`. If `save_channels` is `True` then whether the four separate images should be saved in color or grayscale. If the overall `style` argument is `"grayscale"` then the four CMYK channels will always be `"grayscale"`, no matter what this setting.

Default: `"color"`

### `scale`

Scale of the output image. The maximum output dot diameter is `sample * scale` (which is also the number of possible dot sizes).

Default `1`

### `style`

Either `"color"` or `"grayscale"`. For color, four screens are output, one each for cyan, magenta, yellow and black. For grayscale, only black dots are generated, only the first number in the `angles` list is used, and the `percentage` value is ignored.

Default: `"color"`

## Examples

An example of `make()` using all options:

```python
h.make(
    angles=[15, 75, 0, 45],
    antialias=True,
    filename_addition="_new",
    output_format="jpeg",
    output_quality=95,
    percentage=50,
    sample=5,
    save_channels=True,
    save_channels_format="jpeg",
    save_channels_style="grayscale",
    scale=2,
    style="color"
)
```

See the `examples/` directory for the example images below.

### Original image

![Original image of dog](examples/original.jpg?raw=True)

Other than the `filename_addition` option, the images below have been created
using the options specified.

### Default settings

```python
h.make()
```

![Original image of dog](examples/defaults.jpg?raw=True)

### Custom screen angles

Using different screen angles to reduce moiré (but resulting in a different pattern).

```python
h.make(angles=[15, 45, 0, 75])
```

![Image of dog with custom screen angles](examples/angles.jpg?raw=True)

### Smaller sample

Reducing the sample size and increasing the scale (to increase output detail).

```python
h.make(sample=5, scale=2)
```

![Image of dog with smaller sample size](examples/sample_scale.jpg?raw=True)

### Antialias

With antialiased circles.

```python
h.make(antialias=True)
```

![Antialiased image of dog](examples/antialiased.jpg?raw=True)

### Grayscale

Black and white, setting the angle to 45 (the default angle would be 0, resulting in circles being in rows and columns).

```python
h.make(style="grayscale", angles=[45])
```

![Grayscale image of dog](examples/grayscale.jpg?raw=True)


CHANGELOG.md
# Changelog


## [Unreleased]

* None

## [2.1.1] - 2021-12-21

* Fix error when supplying a list of a single integer for `angles` when using
  `style="grayscale"`.

* Upgrade Pillow requirement to >=8.3.2.

## [2.1.0] - 2021-05-07

* Added `output_format` and `output_quality` options.

* Added `save_channels`, `save_channels_format` and `save_channels_style` options.

* Bug fixes


## [2.0.1] - 2021-05-03

* Upgrade Pillow requirement to >=8.1.2.


## [2.0.0] - 2020-06-20

* Require Python 3 (was python 2.7).
* Upgrade Pillow requirement from 5 to 7
* Add Pipfiles for pipenv
* Add changelog

test.py
import halftone


h = halftone.Halftone("/Users/phil/Desktop/_testpics/cmyk.jpg")
h.make(
    style="color",
    save_channels=True,
    save_channels_style="color",
    save_channels_format="png",
    percentage=100,
    antialias=True,
    output_quality=80,
    output_format="png"

)


halftone.py
import os
import sys

from PIL import Image, ImageDraw, ImageOps, ImageStat

"""
Class: Halftone( path )
Usage:
    import halftone
    h = halftone.Halftone('/path/to/image.jpg')
    h.make()

The bulk of this is taken from this Stack Overflow answer by fraxel:
http://stackoverflow.com/a/10575940/250962
"""


class Halftone(object):
    def __init__(self, path):
        """
        path is the path to the image we want to halftone.
        """
        self.path = path

    def make(
        self,
        sample=10,
        scale=1,
        percentage=0,
        filename_addition="_halftoned",
        angles=[0, 15, 30, 45],
        style="color",
        antialias=False,
        output_format="default",
        output_quality=75,
        save_channels=False,
        save_channels_format="default",
        save_channels_style="color",
    ):
        """
        Leave filename_addition empty to save the image in place.
        Arguments:
            sample: Sample box size from original image, in pixels.
            scale: Max output dot diameter is sample * scale (which is also the
                number of possible dot sizes)
            percentage: How much of the gray component to remove from the CMY
                channels and put in the K channel.
            filename_addition: What to add to the filename (before the extension).
            angles: A list of angles that each screen channel should be rotated by.
                Should be 4 integers when style is 'color', at least 1 for 'grayscale'.
            style: 'color' or 'grayscale'.
            antialias: boolean.
            output_format: "default", "jpeg", "png".
            output_quality: Integer, default 75. Only used when saving jpeg images.
            save_channels: Boolean, whether to save the four separate channel images in
                addition to the main image.
            save_channels_format: "default", "jpeg", "png".
            save_channels_style: "color" or "grayscale".
        """

        self.check_arguments(
            angles=angles,
            antialias=antialias,
            output_format=output_format,
            output_quality=output_quality,
            percentage=percentage,
            sample=sample,
            save_channels=save_channels,
            save_channels_format=save_channels_format,
            save_channels_style=save_channels_style,
            scale=scale,
            style=style,
        )

        f, extension = os.path.splitext(self.path)

        if output_format == "jpeg":
            extension = ".jpg"
        elif output_format.startswith("png"):
            extension = ".png"
        # Else, keep the same as the input file.

        output_filename = "%s%s%s" % (f, str(filename_addition), extension)

        try:
            im = Image.open(self.path)
        except IOError as e:
            raise Exception("Couldn't open source file '%s'" % (self.path)) from e

        if style == "grayscale":
            angles = angles[:1]
            gray_im = im.convert("L")
            channel_images = self.halftone(
                im, gray_im, sample, scale, angles, antialias
            )
            new = channel_images[0]

        else:
            cmyk = self.gcr(im, percentage)
            channel_images = self.halftone(im, cmyk, sample, scale, angles, antialias)

            if save_channels:

                self.save_channel_images(
                    channel_images,
                    channels_style=save_channels_style,
                    channels_format=save_channels_format,
                    output_filename=output_filename,
                    output_quality=output_quality,
                )

            new = Image.merge("CMYK", channel_images)

        if extension == ".jpg":
            new.save(output_filename, "JPEG", subsampling=0, quality=output_quality)
        elif extension == ".png":
            new.convert("RGB").save(output_filename, "PNG")

    def check_arguments(
        self,
        angles,
        antialias,
        output_format,
        output_quality,
        percentage,
        sample,
        save_channels,
        save_channels_format,
        save_channels_style,
        scale,
        style,
    ):
        "Checks all the arguments are valid. Raises TypeError or ValueError if not."

        if not isinstance(angles, list):
            raise TypeError(
                "The angles argument must be a list of 4 integers, not '%s'." % angles
            )

        if style == "grayscale":
            if len(angles) < 1:
                raise ValueError(
                    "The angles argument must be a list of at least 1 integer when "
                    "style is 'grayscale', but it has %s."
                    % len(angles)
                )
        else:
            if len(angles) != 4:
                raise ValueError(
                    "The angles argument must be a list of 4 integers when "
                    "style is 'color', but it has %s."
                    % len(angles)
                )

        for a in angles:
            if not isinstance(a, int):
                raise ValueError(
                    "All elements of the angles list must be integers, "
                    "but it is %s." % angles
                )

        if not isinstance(antialias, bool):
            raise TypeError(
                "The antialias argument must be a boolean, not '%s'." % antialias
            )

        if output_format not in ["default", "jpeg", "png"]:
            raise ValueError(
                "The output_format argument must be one of 'default', "
                "'jpeg' or 'png', not '%s'." % save_channels_format
            )

        if not isinstance(output_quality, int):
            raise TypeError(
                "The output_quality argument must be an integer, not '%s'."
                % output_quality
            )
        if output_quality < 0 or output_quality > 100:
            raise ValueError(
                "The output_quality argument must be between 0 and 100, but it is %s."
                % output_quality
            )

        if not isinstance(percentage, (float, int)):
            raise TypeError(
                "The percentage argument must be an integer or float, not '%s'."
                % percentage
            )

        if not isinstance(sample, int):
            raise TypeError(
                "The sample argument must be an integer, not '%s'." % sample
            )

        if not isinstance(save_channels, bool):
            raise TypeError(
                "The save_channels argument must be a boolean, not '%s'."
                % save_channels
            )

        if save_channels_format not in ["default", "jpeg", "png"]:
            raise ValueError(
                "The save_channels_format argument must be one of 'default', "
                "'jpeg' or 'png', not '%s'." % save_channels_format
            )
        if save_channels_style not in ["color", "grayscale"]:
            raise ValueError(
                "The save_channels_style argument must be one of "
                "'color' or 'grayscale', not '%s'." % save_channels_style
            )

        if not isinstance(scale, int):
            raise TypeError("The scale argument must be an integer, not '%s'." % scale)

        if style not in ["color", "grayscale"]:
            raise ValueError(
                "The style argument must be either 'color' or 'grayscale'."
            )

        return True

    def gcr(self, im, percentage):
        """
        Basic "Gray Component Replacement" function. Returns a CMYK image with
        percentage gray component removed from the CMY channels and put in the
        K channel, ie. for percentage=100, (41, 100, 255, 0) >> (0, 59, 214, 41)
        """
        cmyk_im = im.convert("CMYK")
        if not percentage:
            return cmyk_im
        cmyk_im = cmyk_im.split()
        cmyk = []
        for i in range(4):
            cmyk.append(cmyk_im[i].load())
        for x in range(im.size[0]):
            for y in range(im.size[1]):
                gray = int(
                    min(cmyk[0][x, y], cmyk[1][x, y], cmyk[2][x, y]) * percentage / 100
                )
                for i in range(3):
                    cmyk[i][x, y] = cmyk[i][x, y] - gray
                cmyk[3][x, y] = gray
        return Image.merge("CMYK", cmyk_im)

    def halftone(self, im, cmyk, sample, scale, angles, antialias):
        """
        Returns list of half-tone images for cmyk image. sample (pixels),
        determines the sample box size from the original image. The maximum
        output dot diameter is given by sample * scale (which is also the number
        of possible dot sizes). So sample=1 will presevere the original image
        resolution, but scale must be >1 to allow variation in dot size.
        """

        # If we're antialiasing, we'll multiply the size of the image by this
        # scale while drawing, and then scale it back down again afterwards.
        # Because drawing isn't aliased, so drawing big and scaling back down
        # is the only way to get antialiasing from PIL/Pillow.
        antialias_scale = 4

        if antialias is True:
            scale = scale * antialias_scale

        cmyk = cmyk.split()
        dots = []

        for channel, angle in zip(cmyk, angles):
            channel = channel.rotate(angle, expand=1)
            size = channel.size[0] * scale, channel.size[1] * scale
            half_tone = Image.new("L", size)
            draw = ImageDraw.Draw(half_tone)

            # Cycle through one sample point at a time, drawing a circle for
            # each one:
            for x in range(0, channel.size[0], sample):
                for y in range(0, channel.size[1], sample):

                    # Area we sample to get the level:
                    box = channel.crop((x, y, x + sample, y + sample))

                    # The average level for that box (0-255):
                    mean = ImageStat.Stat(box).mean[0]

                    # The diameter of the circle to draw based on the mean (0-1):
                    diameter = (mean / 255) ** 0.5

                    # Size of the box we'll draw the circle in:
                    box_size = sample * scale

                    # Diameter of circle we'll draw:
                    # If sample=10 and scale=1 then this is (0-10)
                    draw_diameter = diameter * box_size

                    # Position of top-left of box we'll draw the circle in:
                    # x_pos, y_pos = (x * scale), (y * scale)
                    box_x, box_y = (x * scale), (y * scale)

                    # Positioned of top-left and bottom-right of circle:
                    # A maximum-sized circle will have its edges at the edges
                    # of the draw box.
                    x1 = box_x + ((box_size - draw_diameter) / 2)
                    y1 = box_y + ((box_size - draw_diameter) / 2)
                    x2 = x1 + draw_diameter
                    y2 = y1 + draw_diameter

                    draw.ellipse([(x1, y1), (x2, y2)], fill=255)

            half_tone = half_tone.rotate(-angle, expand=1)
            width_half, height_half = half_tone.size

            # Top-left and bottom-right of the image to crop to:
            xx1 = (width_half - im.size[0] * scale) / 2
            yy1 = (height_half - im.size[1] * scale) / 2
            xx2 = xx1 + im.size[0] * scale
            yy2 = yy1 + im.size[1] * scale

            half_tone = half_tone.crop((xx1, yy1, xx2, yy2))

            if antialias is True:
                # Scale it back down to antialias the image.
                w = int((xx2 - xx1) / antialias_scale)
                h = int((yy2 - yy1) / antialias_scale)
                half_tone = half_tone.resize((w, h), resample=Image.LANCZOS)

            dots.append(half_tone)
        return dots

    def save_channel_images(
        self,
        channel_images,
        channels_style,
        channels_format,
        output_filename,
        output_quality,
    ):
        """
        Save the individual CMYK channels as separate images.
        """

        channel_names = (
            ("c", "cyan"),
            ("m", "magenta"),
            ("y", "yellow"),
            ("k", "black"),
        )

        f, extension = os.path.splitext(output_filename)

        if channels_format == "jpeg":
            extension = ".jpg"
        elif channels_format.startswith("png"):
            extension = ".png"
        # Else, keep the same as the input file.

        for count, channel_img in enumerate(channel_images):
            channel_filename = "%s_%s%s" % (
                f,
                channel_names[count][0],
                extension,
            )

            i = ImageOps.invert(channel_img)

            if channels_style == "color" and count < 3:
                i = ImageOps.colorize(i, black=channel_names[count][1], white="white")

            if extension == ".jpg":
                # subsampling=0 seems to make them look crisper.
                i.convert("CMYK").save(
                    channel_filename, "JPEG", subsampling=0, quality=output_quality
                )
            elif extension == ".png":
                i.save(channel_filename, "PNG")


if __name__ == "__main__":

    path = sys.argv[1]

    h = Halftone(path)
    h.make()
